PYME/LMVis/sessionpaths.py

```diff
@@ -0,0 +1,86 @@
+from pathlib import Path
+import os
+from PYME.recipes import base
+from PYME.IO.unifiedIO import is_cluster_uri
+
+import logging
+logger = logging.getLogger(__name__)
+
+SESSIONDIR_TOKEN = '$session_dir$'
+
+# split filename?query type string in fname and query parts
+def parse_fnq(dsfnstr):
+    parts = dsfnstr.split('?')
+    if len(parts) > 1:
+        return  (parts[0],parts[1])# we assume there are no further '?'s in the query string
+    else:
+        return  (parts[0],None)
+
+# join filename and query to form the expected string for session loading
+def fnq_string(fn,query):
+    if query is None:
+        return fn
+    else:
+        return fn + '?' + query
+
+def fnstring_relative(fnstring,sessiondir):
+    fnamep = Path(fnstring)
+    if fnamep.is_absolute():
+        try:
+            fnamerel = str(fnamep.relative_to(sessiondir)) # now make the filenames relative to the session dir path
+        except ValueError: # value error implies it is not possible to make a "clean" relative path, i.e. that the file fnstring is not truely below sessiondir
+            return None # signal via None, we do the further handling in the calling function
+    else:
+        fnamerel = fnstring
+    return fnamerel
+
+def path_is_safe_and_relativeto(fpath,sessiondir):
+    is_safe = (not is_cluster_uri(fpath)) and (not fpath.startswith(SESSIONDIR_TOKEN)) and Path(fpath).is_relative_to(sessiondir)
+    return is_safe
+
+def allpaths_relative_to(session,sessiondir):
+    for ds in session['datasources']:
+        fname,query = parse_fnq(session['datasources'][ds])
+        if not path_is_safe_and_relativeto(fname,sessiondir):
+            return False
+    for module in session['recipe']:
+        [(mn, mod_dict)] = module.items()
+        # we only check registered modules - session loading will anyway fail with an unregistered module
```

Comment:
Is it really possible that there is an un-registered module in the recipe??? If that is the case, is it not better that the session saving fails?